### ISSUE TITLE
.hyper and .race always need one extra item ahead

### ISSUE THREAD

**Issue by AlexDaniel** [rakudo/rakudo#2008]
This code prints nothing and just waits:
```perl6
my $c = Channel.new;
$c.send: 42 for ^1;
for $c.Seq.hyper(:1batch, :1degree) { .say }
```
There is already one item available, so in my opinion it should process it. Change `^1` to `^2` and you will see `42` printed once (even though two items were passed into the channel). 

It works as expected when .hyper is removed.

(BTW, yes, the channel is never closed in the example, and that's ok)

**Comment by AlexDaniel** [comment]
Actually, that seems to be a regression after (2017-06-18) https://github.com/rakudo/rakudo/commit/3d2a521c2558af05cf730d5c230a91237ae1a988

Ping @zoffixznet do you remember what was happening there?

**Comment by zoffixznet** [comment]
IIRC the original commit caused ecosystem regression, it was reverted for release, and after release https://github.com/rakudo/rakudo/commit/3d2a521c2558af05cf730d5c230a91237ae1a988 rever-reverted it.

Presumably the originally identified regression was fixed.

**Comment by jnthn** [comment]
> There is already one item available, so in my opinion it should process it. 

For as long as `hyper` and `race` have been speculated, they've been considered forms of `eager`. In reality, that's been relaxed a bit so one can do streaming things without gobbling memory. However, it's still true in the sense that it can ask for elements ahead of what are currently being processed. (That's how it does batching, and how it can prepare some batches "in advance" to keep the workers busy.)

In this particular case, the reason it's asking for an element ahead is to know whether it's at the end of the sequence, so it can set a `last` flag on the batch of elements. That in turn is needed to know whether to run any `LAST` phasers. That phaser needs to be run in the scope of the final element, so we really need to know about it up front. Also, we can't except very easily detect the absence of `LAST`, and even if we did try to do that, we'd then the same issue would arise the moment somebody added a `LAST` phaser.

We didn't yet implement parallel versions of many operations, so I'd not rule out there being other things that really need a lookahead either.

tl;dr I don't think we can easily get rid of the need for lookahead in the case that currently needs it, and there may be more in the future.

**Comment by AlexDaniel** [comment]
@jnthn I see, thanks.

I'll leave it open for a bit in case someone else wants to comment, but I think this can be closed.

**Comment by ugexe** [comment]
Hmm, wouldn't losing `LAST` inside hyper be a worthy trade-off for not having to know ahead of time that a sequence will contain enough elements to hyper?

**Comment by AlexDaniel** [comment]
Also, what is the meaning of LAST exactly when hypering?

**Comment by zoffixznet** [comment]
> Also, what is the meaning of LAST exactly when hypering?

I remember a rather lengthy discussion of this on IRC, and was hoping someone bothered to actually document its outcome... but I don't see anything [on the phasers page](https://docs.perl6.org/language/phasers#index-entry-Asynchronous_Phasers__LAST-LAST).

**Comment by lizmat** [comment]
Same semantics as with map, afaik

**Comment by zoffixznet** [comment]
```
<Zoffix_> m: hyper for ^10  { LAST .say }
<camelia> rakudo-moar 72ccd432a: OUTPUT: «Phasers in hyper/race not yet implemented. Sorry.␤  in block <unit> at <tmp> line 1␤␤»
```

**Comment by lizmat** [comment]
Closing, while mentioning [`ParaSeq`](https://raku.land/zef:lizmat/ParaSeq)
